master/buildbot/scripts/devproxy.py

```diff
@@ -0,0 +1,168 @@
+# Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+#
+# Copyright Buildbot Team Members
+
+
+import asyncio
+import json
+import logging
+
+import aiohttp  # dev-proxy command requires aiohttp! run 'pip install aiohttp'
+import aiohttp.web
+import jinja2
+
+from buildbot.plugins.db import get_plugins
+
+
+class DevProxy:
+    MAX_CONNECTIONS = 10
+
+    def __init__(self, port, next_url, plugins, unsafe_ssl, auth_cookie):
+        while next_url.endswith('/'):
+            next_url = next_url[:-1]
+        self.next_url = next_url
+        self.app = app = aiohttp.web.Application()
+        self.apps = get_plugins('www', None, load_now=True)
+        self.unsafe_ssl = unsafe_ssl
+        cookies = {}
+        if auth_cookie:
+            if "TWISTED_SESSION" in auth_cookie:  # user pasted the whole document.cookie part!
+                cookies = {k: v for k, v in [c.split("=") for c in auth_cookie.split(";")]}
+                auth_cookie = cookies["TWISTED_SESSION"]
+            cookies = {'TWISTED_SESSION': auth_cookie}
+        logging.basicConfig(level=logging.DEBUG)
+        if plugins is None:
+            plugins = {}
+        else:
+            plugins = json.loads(plugins)
+
+        self.plugins = plugins
+
+        app.router.add_route('*', '/ws', self.ws_handler)
+        for path in ['api', 'auth', 'sse', 'avatar']:
+            app.router.add_route(
+                '*', '/%s{path:.*}' % path, self.proxy_handler)
+        app.router.add_route('*', '/', self.index_handler)
+        for plugin in self.apps.names:
+            if plugin != 'base':
+                staticdir = self.apps.get(plugin).static_dir
+                app.router.add_static('/' + plugin, staticdir)
+        staticdir = self.staticdir = self.apps.get('base').static_dir
+        loader = jinja2.FileSystemLoader(staticdir)
+        self.jinja = jinja2.Environment(
+            loader=loader, undefined=jinja2.StrictUndefined)
+        app.router.add_static('/', staticdir)
+        conn = aiohttp.TCPConnector(
+            limit=self.MAX_CONNECTIONS, verify_ssl=(not self.unsafe_ssl))
+        self.session = aiohttp.ClientSession(connector=conn, trust_env=True, cookies=cookies)
+        self.config = None
+        self.buildbotURL = "http://localhost:{}/".format(port)
+        aiohttp.web.run_app(app, host="localhost", port=port)
+        self.session.close()
+
+    async def ws_handler(self, req):
+        # based on https://github.com/oetiker/aio-reverse-proxy/blob/master/paraview-proxy.py
+        ws_server = aiohttp.web.WebSocketResponse()
+        await ws_server.prepare(req)
+
+        async with self.session.ws_connect(
+            self.next_url + "/ws", headers=req.headers
+        ) as ws_client:
+
+            async def ws_forward(ws_from, ws_to):
+                async for msg in ws_from:
+                    mt = msg.type
+                    md = msg.data
+                    if mt == aiohttp.WSMsgType.TEXT:
```

Comment:
To me it seems that it's not worth extracting `mt` and `md` as the code is more confusing and the space savings are minimal.